### ISSUE TITLE
Sending modules does not really work

### ISSUE THREAD

**Issue by jonasc** [pgiri/dispy#43]
I was not able to correctly send modules consisting of more than just one file to the nodes. I supposed it was due to transferred files being flattened (i.e. no directory hierarchy is preserved). Furthermore this prevents me from sending multiple modules with same (sub-) file names.

I'll publish a PR shortly which addresses this issue.

**Comment by pgiri** [comment]
Can you attach a simple example use case to test this, please?

**Comment by jonasc** [comment]
I'll try to create a simple one. Also note that the pull request only contains code changes for py3 and no py2 (yet).

**Comment by jonasc** [comment]
This example does not work with current master but with my modifications.

**client.py**
~~~python
def compute(x, y, z):
    from a import A, C
    from b import B
    return isinstance(x, A), isinstance(y, B), isinstance(z, C)

if __name__ == '__main__':
    import dispy
    import a
    import b
    import logging
    cluster = dispy.JobCluster(compute, nodes=['*'], depends=[a, b, a.c], loglevel=logging.DEBUG)

    job = cluster.submit(a.A(), b.B(), a.c.C())
    cluster.wait()
    print(job())
    print('Status:', job.status)

    cluster.close()
~~~

**a/__init__.py**
~~~python
from .c import C


class A:
    pass
~~~

**a/c.py**
~~~python
class C:
    pass
~~~

**b/__init__.py**
~~~python
class B:
    pass
~~~

**Comment by jonasc** [comment]
Ok, so I figured out that using my patch together with a scheduler node and SharedJobScheduler results in the following error message on the client:
~~~
2016-07-09 19:13:37 asyncoro - version 4.1 with epoll I/O notifier
2016-07-09 19:13:37 dispy - dispy client at :51347
2016-07-09 19:13:37 dispy - Storing fault recovery information in "_dispy_20160709191337"
2016-07-09 19:13:37 dispy - Pending jobs: 0
2016-07-09 19:13:37 dispy - scheduler quitting (0 / 0)
2016-07-09 19:13:37 dispy - scheduler quit
2016-07-09 19:13:37 dispy - Sending file "/tmp/dp/a/c.py"
2016-07-09 19:13:37 dispy - Sending file "/tmp/dp/b/__init__.py"
2016-07-09 19:13:37 dispy - Sending file "/tmp/dp/a/__init__.py"
Traceback (most recent call last):
  File "client.py", line 11, in <module>
    cluster = dispy.SharedJobCluster(compute, nodes=['*'], depends=[a, b, a.c], scheduler_node='XXXXXX', loglevel=logging.DEBUG)
  File "/YYYYYY/dispy/py3/dispy/__init__.py", line 2591, in __init__
    raise Exception('Computation "%s" could not be sent to scheduler' % self._compute.name)
Exception: Computation "compute" could not be sent to scheduler
2016-07-09 19:13:38 dispy - cluster compute already closed?
2016-07-09 19:13:38 asyncoro - terminating Coro ~timer_task/140108284347112 (daemon)
2016-07-09 19:13:38 asyncoro - terminating Coro ~tcp_server/140108284350304 (daemon)
2016-07-09 19:13:38 asyncoro - AsynCoro terminated
~~~

**Comment by pgiri** [comment]
Let me take a (slightly) different approach to it and submit a patch for you to review.

**Comment by jonasc** [comment]
Ok, I just figured out how to solve the above issue (there was one part in the scheduler I missed).

But I'll gladly look at your proposal.

**Comment by pgiri** [comment]
Attached is patch to send packages.  This turned out to be a bit more complicated than I was expecting. Apparently module.__package__ is not always consistent, so I had to use .__name__ as well if .__package__ is not available (e.g., for 'a.c' in your example). I have not tested the patch thoroughly; likely there are issues with it. Let me know your comments.

[xfer.txt](https://github.com/pgiri/dispy/files/355798/xfer.txt)

**Comment by pgiri** [comment]
I fixed a few issues with the patch above and committed 48004c8. This should be easier than going through xfer.txt attached above.

**Comment by jonasc** [comment]
I guess you mean 8359bb7735bc7c687429d2e7e843762d82021e22, right?
I have not yet tested it but by looking through the code I've encountered some issues I think, namely since the constructor of `_XferFile` now takes a different second parameter. But there are still some instantiations of `_XferFile` that were not changed, e.g. https://github.com/pgiri/dispy/commit/8359bb7735bc7c687429d2e7e843762d82021e22#diff-1eac19ad2aa2da8fd8d2ff8590520c09R2352, https://github.com/pgiri/dispy/commit/8359bb7735bc7c687429d2e7e843762d82021e22#diff-1eac19ad2aa2da8fd8d2ff8590520c09R2819, https://github.com/pgiri/dispy/commit/8359bb7735bc7c687429d2e7e843762d82021e22#diff-8460050624425a840d86f9a84bb9873bR2357, https://github.com/pgiri/dispy/commit/8359bb7735bc7c687429d2e7e843762d82021e22#diff-8460050624425a840d86f9a84bb9873bR2823

**Comment by pgiri** [comment]
Sorry, I messed up my commits for XferFile last night, and had to merge changes in master and then I messed up commit of the merges! Can you please update to current master and let me know if everything works? Thanks.

**Comment by jonasc** [comment]
I seems to work, thanks!

But still there are some occurrences of `_XferFile` which do not correspond to your changes as pointed out in my last comment.

**Comment by pgiri** [comment]
Ah, yes, I see what you mean! Will fix this soon.

**Comment by jonasc** [comment]
Ok, not quite. There seem to be issues still. Take my example from above but change client.py.

**First**
~~~python
# client.py
import a


def compute(o):
    return o

if __name__ == '__main__':
    import dispy
    import logging
    cluster = dispy.SharedJobCluster(compute, nodes=['*'], scheduler_node='XXX', depends=[], loglevel=logging.DEBUG)

    job = cluster.submit(a.C())
    cluster.wait()
    print(job())
    print('Status:', job.status)
    print('Exception:', job.exception)

    cluster.close()
~~~

OK, since I'm passing an `C` object but did not specify it as a dependency I get the following output on the *client* **as expected**:
~~~
None
Status: 9
Exception: Traceback (most recent call last):
  File "/storage/mi/jonascleve/opt/dispy/py3/dispy/dispynode.py", line 192, in _dispy_job_func
    __dispy_job_args = unserialize(__dispy_job_args)
  File "/home/mi/jonascleve/usr/local/lib/python3.5/site-packages/asyncoro/__init__.py", line 76, in unserialize
    return pickle.loads(pkl)
ImportError: No module named 'a'
~~~

**Second**
Now I change `depends=[]` to `depends=[a.C]`. It does not change the above behaviour and **I expected this** since the class carries around it's module's information.

**Third**
Setting it to `depends=[a]` tells me it misses `a.c` so I end up with `depends=[a, a.c]`. I would now expect it to work. And at first it seems to do so, as the (debugging) output on the client is:
~~~
2016-07-10 16:52:15 asyncoro - version 4.1 with epoll I/O notifier
2016-07-10 16:52:15 dispy - dispy client at :51347
2016-07-10 16:52:15 dispy - Storing fault recovery information in "_dispy_20160710165215"
2016-07-10 16:52:15 dispy - Pending jobs: 0
2016-07-10 16:52:15 dispy - scheduler quitting (0 / 0)
2016-07-10 16:52:15 dispy - scheduler quit
2016-07-10 16:52:16 dispy - Sending file "/tmp/dp/a/__init__.py"
2016-07-10 16:52:16 dispy - Sending file "/tmp/dp/a/c.py"
2016-07-10 16:52:16 dispy - Computation compute created with 140590983210152
~~~
But there it just hangs around and does nothing. Looking at the node gives this output:
~~~
2016-07-10 16:52:16 dispynode - New job id 140590997407208 from A.B.C.D/A.B.C.D
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/dispy/py3/dispy/dispynode.py", line 1232, in __reply_Q
    job_reply = self.reply_Q.get()
  File "/usr/local/lib/python3.5/multiprocessing/queues.py", line 113, in get
    return ForkingPickler.loads(res)
ImportError: No module named 'a'
~~~

This is where I do not now further. Thanks for your help.

**Comment by pgiri** [comment]
I committed fixes to XferFile and hopefully it is okay now.

Your example returns object instance. The computation itself worked fine, but when the reply is processed in parent process (that hasn't imported transferred modules) that is not aware of object a.C.  One way to fix this is to define 'setup' function that loads transferred modules. For example, with

    def setup():
        from a import A, C
        return 0

and then adding "'setup'=setup" to JobCluster. However, this only works with JobCluster, not SharedJobCluster, as SharedJobCluster doesn't load modules so won't know about object instances.

Right fix is to serialize objects in replies. For example, with

    def compute(o):
        return serialize(o)

and at the client deserialize reply with `print("Status: %s", dispy.deserialize(job.result))`.

Another fix may be to change cwd in __reply_Q. Let me try that and see if that works.

**Comment by jonasc** [comment]
Ok, your suggestion via serialization works, but let me point out, that the method is called `unserialize` not `deserialize`, just a typo, you might want to fix it in your post to have things clear.

If you find a decent patch that enables one to go without serialization it would be even better.

**Comment by pgiri** [comment]
Thanks for mentioning de/un serialize; fixed the reply above.

Well, I haven't tried it yet, but the solution is to change cwd in '__reply_Q' to current computation's dest_path (where the modules are saved). This should work when only one computation is being served, but won't work if more than one computation is being served simultaneously. I don't think a fix that works in most cases but not in others is a good fix. May be we should just leave it as it is, even if it is inconvenient? If you have ideas / suggestions, let me know.

**Comment by jonasc** [comment]
Is it possible to automate the serialization/deserialization process? I'm thinking about an option, e.g. `auto_serialize` to a `JobCluster` or `SharedJobCluster` which by default is False so that no unnecessary serialization is done.

**Comment by pgiri** [comment]
Well, right after sending reply, I think it may be possible to fix this properly, likely. Right now in __reply_Q, jot replies are queued. Instead, replies can be sent in two bunches. First to send computation ID for which job reply (next item) is so it can change to computation's dest_path and then get the reply, which should work.

However, this still leaves SharedJobCluster case. I don't think loading modules in dispyscheduler is a good idea.

`auto_serialize` option you suggest may be the right thing to do. Alternately, job results can always be serialized at the node and deserialized (`unserialize` should've been 'deserialize`) at the client?

**Comment by pgiri** [comment]
Thinking again about this, I think just serializing/unserializing always (transparent to user) is probably better. The cost in doing this is negligent, as the job_reply (which has result as one attribute) will be serialized anyway. Let me take a stab at this.

**Comment by pgiri** [comment]
I committed (potential) fix to this that serializes results at node and de/un-serializes at client. Tested a few cases.

I also sneaked in couple of other minor tidbits. One is to change 'shutdown' to 'sighandler'. This was something I was playing with yesterday, as it didn't work with Python 2.7: The signal handler was not being called even if I killed process. Digging around, I found that the problem is due to 'event.wait()' in asyncoro. Strangely, doing wait with timeout in a loop with

    while 1:
        if self._complete.wait(1000) == True:
            break

fixes it. I am guessing this is an issue with event wait implementation / semantics in Python 2.7. Will commit asyncoro fix if I don't find a better way to do this. I mention this in case you are curious.

Unless there are any potential issues, I am thinking of releasing both asyncoro and dispy in the next couple of days. Can you tell me how I can attribute your contributions in release announcement? 'jonasc' on github?

**Comment by jonasc** [comment]
Looks all good to me. The attribution is fine this way.

Thanks again for the support and the nice piece of software!

If you don't see anything that's missing here, you can close this issue.

**Comment by pgiri** [comment]
This has been fixed as of 4.6.15 release. Thanks again for your contributions.
